Core.ApplicationServices/SystemUsage/Migration/ItSystemUsageMigrationService.cs

```diff
@@ -0,0 +1,261 @@
+            var unusedSystems = _systemsRepository.GetUnusedSystems(new OrganizationDataQueryParameters(organizationId, queryBreadth, dataAccessLevel));
+
+            //Refine, order and take the amount requested
+            var result = unusedSystems
+                .ByPartOfName(nameContent)
+                .OrderBy(x => x.Name)
+                .Take(numberOfItSystems)
+                .ToList()
+                .AsReadOnly();
+
+            return Result<OperationResult, IReadOnlyList<ItSystem>>.Ok(result);
+        }
+
+        public Result<OperationResult, ItSystemUsageMigration> GetSystemUsageMigration(int usageId, int toSystemId)
+        {
+            if (!CanExecuteMigration())
+            {
+                return Result<OperationResult, ItSystemUsageMigration>.Fail(OperationResult.Forbidden);
+            }
+
+            var itSystemUsage = _itSystemUsageRepository.GetByKey(usageId);
+            if (!_authorizationContext.AllowReads(itSystemUsage))
+            {
+                return Result<OperationResult, ItSystemUsageMigration>.Fail(OperationResult.Forbidden);
+            }
+            var toItSystem = _itSystemRepository.GetByKey(toSystemId);
+            if (!_authorizationContext.AllowReads(toItSystem))
+            {
+                return Result<OperationResult, ItSystemUsageMigration>.Fail(OperationResult.Forbidden);
+            }
+
+            //Map all contract migrations
+            var usageContractIds = itSystemUsage.Contracts.Select(x => x.ItContractId).ToList();
+            var interfaceExhibitUsages = itSystemUsage.ItInterfaceExhibitUsages.ToList();
+            var interfaceUsages = itSystemUsage.ItInterfaceUsages.ToList();
+
+            var contractMigrations = _contractRepository.GetBySystemUsageAssociation(usageId)
+                .AsEnumerable()
+                .Select(contract => CreateContractMigration(interfaceExhibitUsages, interfaceUsages, contract, usageContractIds.Contains(contract.Id)))
+                .ToList()
+                .AsReadOnly();
+
+            return Result<OperationResult, ItSystemUsageMigration>.Ok(
+                new ItSystemUsageMigration(
+                    systemUsage: itSystemUsage,
+                    fromItSystem: itSystemUsage.ItSystem,
+                    toItSystem: toItSystem,
+                    affectedProjects: itSystemUsage.ItProjects,
+                    affectedContracts: contractMigrations));
+        }
+
+        public Result<OperationResult, ItSystemUsage> ExecuteSystemUsageMigration(int usageSystemId, int toSystemId)
+        {
+            using (var transaction = _transactionManager.Begin(IsolationLevel.Serializable))
+            {
+                try
+                {
+                    var migrationConsequences = GetSystemUsageMigration(usageSystemId, toSystemId);
+                    if (migrationConsequences.Status != OperationResult.Ok)
+                    {
+                        transaction.Rollback();
```

Comment:
Er disse rollbacks nødvendige?

Reply:
nej vi skal kun rulle tilbage hvis der f.eks. sker en fejl (exception).. ikke dér for der er jo ikke rigtig nogle ændringer endnu